### ISSUE TITLE
The version control and branching model 

### ISSUE THREAD

**Issue by shafayat1004** [ricebowlq/API_Services-Cloud-Computing#17]
# Version control and Branching model
Since the application is fairly large with a lot of functionality, a lot of developers are going to be working on features and fixes together. This would mean that the branching model should allow for seamless collaboration between them in the codebase but the changes should also be easy to trace. 

In this project, we use a simple and effective branching model that follows these principles:

- **`main` Branch**: This is the branch that contains the stable and production-ready code. It should always be deployable and bug-free. Only authorized developers can merge changes into this branch.

- **`dev` Branch**: This is the branch that contains the code that is under development and testing. It is created from the `main` branch and updated with the latest changes from the `main` branch. It triggers the automatic Continuous Integration and Continuous Delivery (CI/CD) tools that build, test, and deploy the code to the staging environment.

- **Developers would create their own branch from `dev`** based on their name and the features or fixes they might be working on. For example, `dev/alice/login` or `dev/bob/cart`. They can manage their own branch however they see fit, as long as they follow the coding standards and best practices.

- Once they are done with the task, they have to **squash all their commits in their branch into a single commit**. This reduces the clutter and noise in the commit history and makes it easier to review and understand the changes.
- Then they have to **rebase their commit with the latest `dev` branch** and make sure everything works as expected. This ensures that their branch is up-to-date with the latest changes and avoids merge conflicts later on.
- Then they can **submit the branch for Pull Request (PR)**. A PR is a request to merge one branch into another. It allows other developers to review the code, provide feedback, and approve or reject the changes.
- Once the PR is approved and passes all the tests, it can be **merged back into the `dev` branch**. This creates a clean linear history of the changes in the codebase, which makes it easier to track and debug issues.
- **When the `dev` branch is ready for release, it can be merged into the `main` branch** and tagged with a version number.

By following this branching model, we can achieve a high level of collaboration and quality in our project. It allows us to work on different features independently, without interfering with each other’s work. It also ensures that our code is always consistent and reliable, as we only merge changes that are tested, reviewed, and deployed.
